### ISSUE TITLE
Cross-Chunk Slicing Memory Error

### ISSUE THREAD

**Issue by trexfeathers** [dask/dask#5540]
My team's userbase of scientists often need to slice data in shapes that result in each slice touching high numbers of the source data's chunks. This causes sub-optimal memory usage by Dask - see [Specifying Chunk shapes](https://docs.dask.org/en/latest/array-chunks.html#specifying-chunk-shapes) in the documentation. In some examples this is causing memory errors, and unfortunately the users have no control over the source data's chunking. The current difficulty with 'cross-chunk slicing' is therefore a blocker preventing this group of scientists from adopting Dask.

Could changes be made to Dask that would remove this limitation, without negatively affecting performance in other scenarios? I see that this has been discussed in #3595, but that no conclusions have been reached since workarounds were possible. We are hoping a general solution may be reachable since developing case-by-case workarounds for each of our users is beyond the resource we have.

I'd be happy to spend time upskilling enough to discuss / work on a solution myself, but first it would be very helpful to know whether this might be a dead-end or whether it is worth investing the time. Thanks!

_The below two examples of ‘cross-chunk slicing’ cause memory errors with Dask 2.5.2, Numpy 1.17.2._

<details>
    <summary>System specs</summary>

````
$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz
Stepping:              1
CPU MHz:               2599.998
BogoMIPS:              5199.99
Virtualization:        VT-x
Hypervisor vendor:     VMware
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              35840K
NUMA node0 CPU(s):     0-3
````
```
$ lsmem
RANGE                                 SIZE  STATE REMOVABLE BLOCK
0x0000000000000000-0x00000000bfffffff   3G online        no  0-23
0x0000000100000000-0x00000001bfffffff   3G online        no 32-55

Memory block size:       128M
Total online memory:       6G
Total offline memory:      0B

```

</details>

```python
# Example using the shape of a real dataset that has this problem.
shape_original = da.random.random(
    size=[11280, 720, 1440],
    chunks=[16, -1, -1])
slice_original = shape_original[:, :71, :130]
slice_original.compute()
```
```python
# Simplified example demonstrating the same problem.
shape_simplified = da.random.random(
    size=[100000, 100000],
    chunks=[1, -1])
slice_simplified = shape_simplified[:, 1]
slice_simplified.compute()
```

**Comment by jrbourbeau** [comment]
Thanks for opening up an issue @trexfeathers. Do you have any thoughts on what a general solution may look like? That could be useful in helping guide any further discussion. You may also consider commenting on #3595 to see if others on that thread have made any progress since the original conversation took place

**Comment by mrocklin** [comment]
One grand solution to this problem is to have a higher level numpy-like library that could capture all of the users operations, and then optimize those calls accordingly for things like chunk structure.  Unfortunately, dask array doesn't have any mechanism for something like this today.  There has been some talk about doing some lightweight version of this with the new [high level graphs](https://docs.dask.org/en/latest/high-level-graphs.html), but even these probably aren't going to be able to solve the larger underlying problem here.

**Comment by mrocklin** [comment]
If I had a person to work on this problem for six months I would suggest that they make another numpy-like library that captured a call graph of all of the operations, and then a bunch of routines to optimize around that structure.  Users would use that, and that would then lower down to dask array.  

At one point I made a small prototype of this, but I can't find it now.  I imagine it's on some github repository somewhere.  It probably isn't of much value though.

**Comment by trexfeathers** [comment]
Thanks for your contributions everyone. Knowing the potential scale of the solution has helped my team make time for writing individual workarounds, this being much quicker than understanding and solving the root problem. We've also been able to spread awareness of the situation to colleagues who might encounter something similar.

If the individual workarounds become too much we may revisit and attempt to put together a PR for Dask, but we are happy for now. Feel free to close this issue if appropriate.

**Comment by dcherian** [comment]
@trexfeathers Could you summarize what workarounds you ended up using?

**Comment by trexfeathers** [comment]
> @trexfeathers Could you summarize what workarounds you ended up using?

@pp-mo care to share?

**Comment by pp-mo** [comment]
>  what workarounds you ended up using?

Firstly, the key performance problem was really simple to fix by replacing the `array.compute()` with :
```
def compute_via_store(data):
    result = np.empty(data.shape, dtype=data.dtype)
    da.store(data, result)
    return result
```

Unfortunately, this wasn't appropriate to the needs of a user creating a generic processing library, because they needed to deliver a dask-array result that could be further indexed, and the compute phase is down to the end-user.

So then I produced a "safe-slice" routine.  
Notes:
  * replaces simple slicing op with a equivalent function that constructs an equivalent-valued lazy   * scheme:
    * pick apart the array into its existing chunks 
    * slice each chunk independently, using indexing derived from the 'top level' one
result, but computing a bit differently
    * make an "np.copy" of each indexed chunk
      * **this bit is crucial** : [as suggested here](https://github.com/dask/dask/issues/3595#issuecomment-449546228)
  * tried to generalise the indexing operation, but it is complex
  * does not assume structure of the underlying array
  * does not rely too much on dask internals, but does unpick the chunks data
  * it does work, but ..
    * only a hacky POC
    * specific to the indexing operation
    * applies exclusively to the case where the slicing, *as applied to each chunk*, results in a small fraction of the total data

documented as :
```
def safeslice(data, indices, chunks=None):
    """
    Return a subset of a dask array, but with indexing applied independently to
    each slice of the input array, *prior* to their recombination to produce
    the result array.

    Args:

    * data (dask array):
        input data
    * indices (int or slice or tuple(int or slice)):
        required sub-section of the data.

    Kwargs:

    * chunks (list of (int or "auto")):
        chunking argument for 'rechunk' applied to the input.
        If set, forces the input to be rechunked as specified.
        ( This replaces the normal operation, which is to rechunk the input
        making the indexed dimensions undivided ).
        Mainly for testing on small arrays.

    .. note::

        'indices' currently does not support Ellipsis or newaxis.

    """
```

Here's the actual source :
[safe_cube_slice.py.txt](https://github.com/dask/dask/files/4354338/safe_cube_slice.py.txt)

**Comment by dcherian** [comment]
@pp-mo Thanks you for taking the time to write that up! 
It will take me some time to fully understand so I may have some questions for you later.
